### PR TITLE
Implement query algorithm to find available meeting times

### PR DESCRIPTION
**Pull request by BryanTriana**

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,116 @@
...
+  /**
+   * Finds the available times in a day in which all meeting participants can attend.
+   *
+   * @param events The events occurring in a single day
+   * @param request The meeting that needs to be accommodated
+   * @return List of all available meeting times within a single day
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:35]: might be good to add to the documentation that you don't support requests that are longer than a day

@param request the meeting that needs to be accommodated - can't be longer than a day

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,116 @@
...
+import java.util.List;
 
+/**
+ *  Holds utility method 'query' to find the available times for a meeting
+ */
 public final class FindMeetingQuery {
+  /**
+   * Finds the available times in a day in which all meeting participants can attend.
+   *
+   * @param events The events occurring in a single day
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:30]: java style nit: for block comments, like @param and @return, we don't capitalize

@param events the events occurring in a single day


(confusing, I know)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,116 @@
...
+      List<TimeRange> availableTimes = new ArrayList<>();
+      availableTimes.add(TimeRange.WHOLE_DAY);
+
+      return availableTimes;
+    }
+
+    Collections.sort(conflictingIntervals, TimeRange.ORDER_BY_START);
+
+    List<TimeRange> mergedIntervals = getMergedIntervals(conflictingIntervals);
+    List<TimeRange> availableTimes = new ArrayList<>();
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: can be initialized once around line 46

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,116 @@
...
+
+    int availableDuration = mergedIntervals.get(0).start() - TimeRange.START_OF_DAY;
+
+    if (availableDuration >= request.getDuration()) {
+      availableTimes.add(TimeRange.fromStartDuration(TimeRange.START_OF_DAY, availableDuration));
+    }
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:63]: can you add some documentation about what you're doing here?

**Comment by BryanTriana** [comment]
I made a new PR with changes from another branch that included optional attendees as well as new tests.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,116 @@
...
+
+    for (int i = 0; i < mergedIntervals.size() - 1; ++i) {
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:65]: super tiny nit: usually we do i++ (even though yours is also correct)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,116 @@
...
+      availableDuration = mergedIntervals.get(i + 1).start() - mergedIntervals.get(i).end();
+
+      if (availableDuration >= request.getDuration()) {
+        availableTimes.add(
+            TimeRange.fromStartDuration(mergedIntervals.get(i).end(), availableDuration));
+      }
+    }
+
+    availableDuration =
+        TimeRange.END_OF_DAY + 1 - mergedIntervals.get(mergedIntervals.size() - 1).end();
```
Comment (rharagutchi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: also documentation here; overall something like

// Look for available times at the start of the day until the first event.
// Look for available times in between existing events.
// .... etc

**Pull request closed without merging**